crates/sui-core/src/authority.rs

```diff
@@ -3902,6 +3910,82 @@ impl AuthorityState {
         Ok(new_epoch_store)
     }
 
+    pub async fn settle_transactions_for_testing(
+        &self,
+        ckpt_seq: CheckpointSequenceNumber,
+        effects: &[TransactionEffects],
+    ) {
+        let builder = AccumulatorSettlementTxBuilder::new(
+            Some(self.get_transaction_cache_reader().as_ref()),
+            effects,
+            ckpt_seq,
+            0,
+        );
+        let epoch_store = self.epoch_store_for_testing();
+        let epoch = epoch_store.epoch();
+        let (settlements, barrier) = builder.build_tx(
+            epoch_store.protocol_config(),
+            epoch,
+            epoch_store
+                .epoch_start_config()
+                .accumulator_root_obj_initial_shared_version()
+                .unwrap(),
+            ckpt_seq,
+            ckpt_seq,
+        );
+
+        let settlements: Vec<_> = settlements
+            .into_iter()
+            .map(|tx| {
+                VerifiedExecutableTransaction::new_system(
+                    VerifiedTransaction::new_system_transaction(tx),
+                    epoch,
+                )
+            })
+            .collect();
+        let barrier = VerifiedExecutableTransaction::new_system(
+            VerifiedTransaction::new_system_transaction(barrier),
+            epoch,
+        );
+
+        let assigned_versions = epoch_store
+            .assign_shared_object_versions_for_tests(
+                self.get_object_cache_reader().as_ref(),
+                &settlements,
+            )
+            .unwrap();
+
+        let version_map = assigned_versions.into_map();
+
+        for tx in settlements {
+            let env = ExecutionEnv::new()
+                .with_assigned_versions(version_map.get(&tx.key()).unwrap().clone());
+            let (effects, _) = self
+                .try_execute_immediately(&tx, env, &epoch_store)
+                .await
+                .unwrap();
+            assert!(effects.status().is_ok());
+        }
+
+        let assigned_versions = epoch_store
```

Comment:
Could we just merge barrier into settlements and then assign version / execute together?